### ISSUE TITLE
Help

### ISSUE THREAD

**Issue by juanpablobr** [google/rekall#164]
Hi guys, I am learning memory analysis, and I've been trying to resolve the volatilty examples images using rekall.

I have some problems right now:
- What is the equivalent in rekall to the vadinfo volatility plugin. I am exactly looking for this kind of output.

Pid: 4
VAD node @823c4200 Start 00010000 End 00033fff Tag Vad
Flags:
Commit Charge: 0 Protection: 4
ControlArea @823c1ac0 Segment e100f690
Dereference list: Flink 00000000, Blink 00000000
NumberOfSectionReferences: 1 NumberOfPfnReferences: 0
NumberOfMappedViews: 25 NumberOfUserReferences: 25
WaitingForDeletion Event: 00000000
Flags: Commit, HadUserReference
**FileObject: none**
First prototype PTE: e100f6d0 Last contiguous PTE: e100f7e8
Flags2: Inherit
File offset: 00000000 

- How can I extend the output of the threads command, to ge something like this?

ETHREAD: 0x822ee3c8 Pid: 1956 Tid: 1984
Tags:
Created: 2011-04-10 21:05:29
Exited: -
Owning Process: 0x8227bb28 'EXPLORER.EXE'
Attached Process: 0x8227bb28 'EXPLORER.EXE'
State: Waiting:UserRequest
BasePriority: 0x8
Priority: 0x8 ...

- How can I use the strings utility with Rekall.

Thanks so much in advance.

**Comment by scudette** [comment]
On 29/07/2016, juanpablobr <notifications@github.com> wrote:
> Hi guys, I am learning memory analysis, and I've been trying to resolve the
> volatilty examples images using rekall.

Hi, welcome to the list.

> I have some problems right now:
> - What is the equivalent in rekall to the vadinfo volatility plugin. I am
> exactly looking for this kind of output.
>
> Pid: 4
> VAD node @823c4200 Start 00010000 End 00033fff Tag Vad
> Flags:
> Commit Charge: 0 Protection: 4
> ControlArea @823c1ac0 Segment e100f690
> Dereference list: Flink 00000000, Blink 00000000
> NumberOfSectionReferences: 1 NumberOfPfnReferences: 0
> NumberOfMappedViews: 25 NumberOfUserReferences: 25
> WaitingForDeletion Event: 00000000
> Flags: Commit, HadUserReference
> **FileObject: none**
> First prototype PTE: e100f6d0 Last contiguous PTE: e100f7e8
> Flags2: Inherit
> File offset: 00000000

The equivalent is the "vad" plugin which has normally more concise and
readable output. It does not by default show much of the extra details
you show above (its rarely needed) but you can always see it if you
really want by writing a simple console script:

plugin = session.plugins.vad()
for row in plugin.collect():
   process = row[0]
   mmvad = row[2]
   print process.name, mmvad.ControlArea

> - How can I extend the output of the threads command, to ge something like
> this?

The threads plugin already shows a lot of the most frequently needed
fields and you can use a similar technique to show more fields. Rekall
plugins are structured which means they usually produce tables with a
fixed number of columns. We try to only show the most useful fields
and declutter the output.

> - How can I use the strings utility with Rekall.

If you have a raw image you can just run strings on the image and
resolve the physical offset to a virtual address (process owner). For
example say you found a hit at offset 0x20e21000 in the raw image
(physical address space) for an mz header:

[1] win7.elf 00:57:21> dump 0x20e21000, "P"
    Offset                                   Data
                          Comment
-------------- -----------------------------------------------------------------
----------------------------------------
    0x20e21000 4d 5a 90 00 03 00 00 00 04 00 00 00 ff ff 00 00
MZ..............
    0x20e21010 b8 00 00 00 00 00 00 00 40 00 00 00 00 00 00 00
........@.......
    0x20e21020 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
................
    0x20e21030 00 00 00 00 00 00 00 00 00 00 00 00 e0 00 00 00
................
    0x20e21040 0e 1f ba 0e 00 b4 09 cd 21 b8 01 4c cd 21 54 68
........!..L.!Th
    0x20e21050 69 73 20 70 72 6f 67 72 61 6d 20 63 61 6e 6e 6f
is.program.canno
    0x20e21060 74 20 62 65 20 72 75 6e 20 69 6e 20 44 4f 53 20
t.be.run.in.DOS.

You could just ask - which process owns this string? where did it come
from? (use the ptov plugin - physical to virtual)
[1] win7.elf 00:58:28> ptov 0x20e21000
File Mapping (C:\Windows\System32\kernel32.dll @ 0x0
Mapped in 0xfa80024f85d0 svchost.exe   236 @ 0x771f0000
Mapped in 0xfa8002259060 csrss.exe   348 @ 0x771f0000
Mapped in 0xfa8000901060 wininit.exe   384 @ 0x771f0000
Mapped in 0xfa8000900420 csrss.exe   396 @ 0x771f0000
Mapped in 0xfa8002282710 winlogon.exe   436 @ 0x771f0000
Mapped in 0xfa800206d5f0 services.exe   480 @ 0x771f0000
Mapped in 0xfa800183ab30 lsass.exe   496 @ 0x771f0000
Mapped in 0xfa800239db30 lsm.exe   504 @ 0x771f0000
Mapped in 0xfa80028a1640 WmiPrvSE.exe   592 @ 0x771f0000

Its showing you that this MZ header comes from the file kernel32.dll
and it is mapped into all these processes at that offset.

Usually you only use strings if you dont know what you are looking for
- if you do then use yara scan to just find it. You can run strings
first to find interesting strings and then yarascan to look for them.

Hope this helps,
Michael.

**Comment by juanpablobr** [comment]
Thanks for your answer Michael. What do you mean with "console script", can I use a script with rekall in interactive mode?

**Comment by scudette** [comment]
Console scripts are described here:
http://www.rekall-forensic.com/docs/Manual/tutorial.html#_automating_rekall
5.4. Example: Automating the Rekall Console

and are the easiest way to automate Rekall. You basically write anything you would type in the interactive console in an external file and then you just run it with `run -i myfile.py`. The script will run just as if you typed it in the console. You can do anything with that script (eg write to files, read input etc.). 

Basically unless what you want to write can be easily reused in other contexts you are better off to write a script rather than a plugin. If you think you would like to contribute the code and it would be widely useful then you can write a proper plugin.

**Comment by juanpablobr** [comment]
I already understood the difference between strings and yara. What I am trying to do right now is to use a regex to find in memory all coincidences of URLs in the offset space of some suspicions process.

I have this regex "/[a-zA-Z0-9\-\.]+\.(com|ru|org|net|mil|edu|biz|name|info)/", but I am not able to make it work in the yarascan or simple_yarascan plugin.

I've been trying the following arguments with not sucess:

yarascan(string='/[a-zA-Z0-9\-\.]+\.(com|org|net|mil|edu|biz|name|info)/',eprocess=0x820618c8, scan_physical=1)

yarascan(yara_expression=='/[a-zA-Z0-9\-\.]+\.(com|org|net|mil|edu|biz|name|info)/',eprocess=0x820618c8, scan_physical=1)

What I am doing wrong? If i search for a simple string, e.g. "http" it works, but is not what I need right now.

thanks in advance

**Comment by scudette** [comment]
The yara scan plugins provide the string arg as a shortcut (most of the time you just want to grep for a string). But if you want to run a proper yara rule you need to give it as yara_expression. You can look at the yara manual (http://yara.readthedocs.io/en/v3.5.0/) for the exact format of a yara rule but something like this might work:

```
$ rule = """
rule foobar {
  strings:
       $ = /[a-zA-Z0-9-.]+.(com|org|net|mil|edu|biz|name|info)/

  conditions:
       any of them
}"""

yarascan(yara_expression=rule, scan_physical=1)

**Comment by juanpablobr** [comment]
Thank you Michael. With your advice I found a yara rule to search for urls.
